### PR TITLE
add check that vnic with same name doesn't has the same network

### LINKED ISSUE #82: Simulate VNIC name rules
There cannot be two VNICs on the same network with the same name. The mock implementation should reflect this.

### PR DESCRIPTION
**Pull request by mgold1234**
## Please describe the change you are making

This PR fixes #82 and check that vnic doesn't has same network

## Are you the owner of the code you are sending in, or do you have permission of the owner?

yes

## The code will be published under the BSD 3 clause license. Have you read and understood this license?

yes

### CODE REVIEW

**Review by janosdebugs — CHANGES_REQUESTED** [review]
Hey @mgold1234 thank you for your continued work on this. Can I ask you to add tests for both positive and negative for this case? It would be useful to verify the behavior against the real oVirt engine as well as the mock facility.

**Comment by mgold1234** [comment]
> Hey @mgold1234 thank you for your continued work on this. Can I ask you to add tests for both positive and negative for this case? It would be useful to verify the behavior against the real oVirt engine as well as the mock facility.

hi @janosdebugs, I added test that check different network (if there is exist one, else check the same network)  and same name

File: client_nic_test.go [hunk 1]
```diff
@@ -47,6 +47,19 @@ func assertCannotCreateNIC(
 	}
 }
 
+func assertCannotCreateNICWithSameNameDiffNetwork(
```
Comment (janosdebugs) [inline client_nic_test.go:50]: I would suggest renaming this to `assertCannotCreateNICWithVNICProfile` since that more accurately covers what this function does and makes the function reusable.

File: client_nic_test.go [hunk 2]
```diff
@@ -47,6 +47,19 @@ func assertCannotCreateNIC(
...
+	t *testing.T,
+	vm ovirtclient.VM,
+	name string,
+	diffVNICProfile string,
+	params ovirtclient.BuildableNICParameters,
+) {
+	nic, _ := vm.CreateNIC(name, diffVNICProfile, params)
+	if nic != nil {
+		t.Fatalf("create 2 NICs with same name %s", name)
```
Comment (janosdebugs) [inline client_nic_test.go:59]: Please also change this message.

File: client_nic_test.go [hunk 3]
```diff
@@ -47,6 +47,19 @@ func assertCannotCreateNIC(
...
 
+func assertCannotCreateNICWithSameNameDiffNetwork(
+	t *testing.T,
+	vm ovirtclient.VM,
+	name string,
+	diffVNICProfile string,
+	params ovirtclient.BuildableNICParameters,
+) {
+	nic, _ := vm.CreateNIC(name, diffVNICProfile, params)
+	if nic != nil {
```
Comment (janosdebugs) [inline client_nic_test.go:58]: It is preferable from a readability perspective to check the error, not the returned object.

File: client_nic_test.go [hunk 4]
```diff
@@ -34,7 +34,7 @@ func assertCanCreateNIC(
 	return nic
 }
 
-func assertCannotCreateNIC(
+func assertCannotCreateNICWithSameName(
```
Comment (janosdebugs) [inline client_nic_test.go:37]: This name change is misleading. The error message below should also be changed.

File: client_nic_create_test.go [hunk 5]
```diff
@@ -58,3 +58,34 @@ func TestDuplicateVMNICCreation(t *testing.T) {
...
+	)
+	assertNICCount(t, vm, 0)
+	nic1 := assertCanCreateNIC(
+		t,
+		helper,
+		vm,
+		nicName,
+		ovirtclient.CreateNICParams())
+	assertNICCount(t, vm, 1)
+	DiffVNICProfile, _ := assertCanFindDiffVNICProfile(helper, helper.GetVNICProfileID())
```
Comment (janosdebugs) [inline client_nic_create_test.go:81]: What happens if there is no second vnic profile? We cannot assume there is a secondary profile available. Can we create a second profile? What if the secondary profile is on the same network as the primary? Does it still work?

**Review by janosdebugs — CHANGES_REQUESTED** [review]
Hey @mgold1234 thank you so much for working on this. I know this is a tricky issue, I commented on a few corner cases that we may want to check on. Could you try and get those covered too? Thank you.

File: client_nic_test.go (hunk 1 above)
Comment (mgold1234) [reply client_nic_test.go:50]: WDYT about assertCannotCreateNICWithSameVNICProfile? 
it is not clear just with vnic profile.

File: client_nic_test.go (hunk 2 above)
Comment (mgold1234) [reply client_nic_test.go:59]: done

File: client_nic_test.go (hunk 3 above)
Comment (mgold1234) [reply client_nic_test.go:58]: done

File: client_nic_test.go (hunk 1 above)
Comment (janosdebugs) [reply client_nic_test.go:50]: We could do that, however, we are trying to keep the assert functions reusable in other test cases. The naming should convey what the assert functions are doing rather than the current use case so future contributors feel that they can safely reuse the assert function. If  the assert function alone does not convey the logic in the test case using the assert function we can always add comments explaining what's happening.

Comment (mgold1234) [reply client_nic_test.go:50]: NP

File: client_nic_test.go (hunk 4 above)
Comment (mgold1234) [reply client_nic_test.go:37]: WDYT about -("create 2 NICs with same name %s (%v)", name, err)

Comment (mgold1234) [reply client_nic_test.go:37]: @janosdebugs , we have some functions for create nic:
1)  assertCanCreateNIC
2) assertCannotCreateNICWithSameName
3) assertCannotCreateNICWithVNICProfile
and more...

regarding 2 and 3 what do you suggest? if I change 2 as it was assertCannotCreateNIC, what is the differerent from 3?
maybe change it to 'assertCannotCreateNICWithName'? 
and 3 will be assertCannotCreateNICWithVNICProfile?

File: client_nic_create_test.go (hunk 5 above)
Comment (mgold1234) [reply client_nic_create_test.go:81]: @janosdebugs , so I will change the function to catch the case that if there is no secondary profile then we will check the assertCannotCreateNICWithSameName and if we have more then 1 profile we will check this -assertCannotCreateNICWithVNICProfile.
WDYT?

Comment (janosdebugs) [reply client_nic_create_test.go:81]: Yes. Ideally, we'll skip the test if there is no secondary profile available. Later we can create a secondary profile to test.

File: client_nic_create_test.go [hunk 6]
```diff
@@ -58,3 +59,72 @@ func TestDuplicateVMNICCreation(t *testing.T) {
...
+	)
+	assertNICCount(t, vm, 0)
+	nic1 := assertCanCreateNIC(
+		t,
+		helper,
+		vm,
+		nicName,
+		ovirtclient.CreateNICParams())
+	assertNICCount(t, vm, 1)
+	DiffVNICProfileDiffNetwork, _ := assertCanFindDiffVNICProfileDiffNetwork(helper, helper.GetVNICProfileID())
```
Comment (janosdebugs) [inline client_nic_create_test.go:81]: In general, please use the error return and don't ignore that parameter. In this case, however, `t.Skipf` should be used in the assert function.

In Go, error checking should always rely on the returned `error` parameter to make the code more readable and less error-prone.

File: client_nic_create_test.go [hunk 7]
```diff
@@ -58,3 +59,72 @@ func TestDuplicateVMNICCreation(t *testing.T) {
...
+	if DiffVNICProfileDiffNetwork == "" {
```
Comment (janosdebugs) [inline client_nic_create_test.go:82]: This condition should not be in this test case. If we don't find a second network we can use we should simply skip the test, not try to do something else.

File: client_nic_create_test.go [hunk 8]
```diff
@@ -58,3 +59,72 @@ func TestDuplicateVMNICCreation(t *testing.T) {
...
+	} else {
+		assertCannotCreateNICWithVNICProfile(
+			t,
+			vm,
+			nicName,
+			DiffVNICProfileDiffNetwork,
+			ovirtclient.CreateNICParams())
+	}
+	assertNICCount(t, vm, 1)
+	assertCanRemoveNIC(t, nic1)
```
Comment (janosdebugs) [inline client_nic_create_test.go:98]: Removing the NIC is not necessary. When the VM is deleted at the end of the test the NIC is automatically destroyed.

File: client_nic_create_test.go [hunk 9]
```diff
@@ -58,3 +59,72 @@ func TestDuplicateVMNICCreation(t *testing.T) {
...
+	)
+	assertNICCount(t, vm, 0)
+	nic1 := assertCanCreateNIC(
+		t,
+		helper,
+		vm,
+		nicName,
+		ovirtclient.CreateNICParams())
+	assertNICCount(t, vm, 1)
+	DiffVNICProfileSameNetwork, _ := assertCanFindDiffVNICProfileSameNetwork(t, helper, helper.GetVNICProfileID())
```
Comment (janosdebugs) [inline client_nic_create_test.go:120]: Please make variables lower case. Only exported variables made available to other packages should start with an uppercase letter.

File: client_vnicprofile_test.go [hunk 10]
```diff
@@ -55,3 +55,38 @@ func assertCanCreateVNICProfile(t *testing.T, helper ovirtclient.TestHelper) ovi
...
+func assertCanFindDiffVNICProfileDiffNetwork(helper ovirtclient.TestHelper, vnicProfileID string) (string, error) {
+	client := helper.GetClient()
+	existVNICProfile, err := client.GetVNICProfile(vnicProfileID)
+	if err != nil {
+		return "", fmt.Errorf("failed to verify VNIC profile ID %s", vnicProfileID)
+	}
+	networkID := existVNICProfile.NetworkID()
+	vnicProfiles, err := client.ListVNICProfiles()
+	if err != nil {
+		return "", fmt.Errorf("failed to list VNIC profiles (%w)", err)
```
Comment (janosdebugs) [inline client_vnicprofile_test.go:68]: Please use `t.Fatalf`.

File: client_vnicprofile_test.go [hunk 11]
```diff
@@ -55,3 +55,38 @@ func assertCanCreateVNICProfile(t *testing.T, helper ovirtclient.TestHelper) ovi
...
+	}
+	for _, vnicProfile := range vnicProfiles {
+		vnicID := vnicProfile.ID()
+		if vnicID != vnicProfileID && vnicProfile.NetworkID() != networkID {
+			return vnicID, nil
+		}
+	}
+	return "", fmt.Errorf(
```
Comment (janosdebugs) [inline client_vnicprofile_test.go:76]: In this case we should use `t.Skipf` since the conditions required by the test are not met so we can't perform the test.

File: client_vnicprofile_test.go [hunk 12]
```diff
@@ -55,3 +55,38 @@ func assertCanCreateVNICProfile(t *testing.T, helper ovirtclient.TestHelper) ovi
...
+		"failed to find different VNIC profile ID and Different Network for testing, use the exiting one")
+}
+
+func assertCanFindDiffVNICProfileSameNetwork(t *testing.T, helper ovirtclient.TestHelper, vnicProfileID string) (string, error) {
+	client := helper.GetClient()
+	newVNICProfile := assertCanCreateVNICProfile(t, helper)
+	existVNICProfile, err := client.GetVNICProfile(vnicProfileID)
+	if err != nil {
+		return "", fmt.Errorf("failed to verify VNIC profile ID %s", vnicProfileID)
```
Comment (janosdebugs) [inline client_vnicprofile_test.go:85]: Please use `t.Fatalf` in assert functions to make using them easier.

File: client_vnicprofile_test.go [hunk 13]
```diff
@@ -55,3 +55,38 @@ func assertCanCreateVNICProfile(t *testing.T, helper ovirtclient.TestHelper) ovi
...
+	}
+	if newVNICProfile.NetworkID() == existVNICProfile.NetworkID() {
+		return newVNICProfile.ID(), nil
+	}
+	return "", fmt.Errorf(
```
Comment (janosdebugs) [inline client_vnicprofile_test.go:90]: Here we may use `t.Skipf` to make sure we skip the test if the preconditions are not met.

File: mock_nic_create.go [hunk 14]
```diff
@@ -22,7 +22,7 @@ func (m *mockClient) CreateNIC(
 	}
 	for _, n := range m.nics {
 		if n.name == name {
```
Comment (janosdebugs) [inline mock_nic_create.go:24]: This does not seem to implement the naming rules accurately. It will fail on all name conflicts, not just when it's the same network/vnic profile ID. If the tests pass on the mock then the tests are not working correctly.

**Review by janosdebugs — CHANGES_REQUESTED** [review]
Hey @mgold1234 thank you very much for your extensive work on this PR. There are several things to look at. The main points:

- Assert functions in this code base should use `t.Fatalf` to fail or `t.Skipf` to skip a test. We skip tests if the preconditions are not met (e.g. a second network is not available), while we fail a test when an actual error happened (e.g. the engine didn't respond).
- In Go please always check the returned `error` type instead of the return value. This makes the code more readable and less likely to introduce bugs.
- It seems the mock implementation does not correctly implement the rules, it will fail if there is any name conflict. Since the test passes it means that there is at least one test missing. There should be a failing test on the current implementation. This test should check that you *can* create two NICs with the same name on different VNIC profiles or different networks (depending on your investigation).

Since we don't know what the rules are, depending on your investigation of the naming conflict you need to implement *one* of the following two cases with 3 tests:

If the naming conflict applies to VNIC profiles:

- I cannot create two NICs with the same name on the VNIC profile.
- I can create two NICs with the same name on a different VNIC profile but the same network.
- I can create two NICs with the same name on different networks.

If the naming conflict applies to networks:

- I cannot create two NICs with the same name on the VNIC profile (which ever applies).
- I cannot create two NICs with the same name on a different VNIC profile but the same network.
- I can create two NICs with the same name on different networks.

File: mock_nic_create.go [hunk 15]
```diff
@@ -21,8 +21,8 @@ func (m *mockClient) CreateNIC(
 		return nil, newError(ENotFound, "VM with ID %s not found for NIC creation", vmid)
 	}
 	for _, n := range m.nics {
-		if n.name == name {
-			return nil, newError(ENotFound, "NIC with name %s is already in use", name)
+		if n.name == name && m.vnicProfiles[n.vnicProfileID].networkID == m.vnicProfiles[vnicProfileID].networkID {
```
Comment (janosdebugs) [inline mock_nic_create.go:24]: I checked in my local setup, if I remove this change the test still passes. That means that there is no test that catches this bug.

**Review by janosdebugs — CHANGES_REQUESTED** [review]
Hey @mgold1234 as indicated in the previous review, the test needs to fail if the correct implementation is removed. This is not the case, when removing the latest change the test still passes. Please add a test that tests this case.

**Comment by mgold1234** [comment]
@janosdebugs , there is an option that I can see the result of run here, before you do the review? 
there is a way to skip on workflows awaiting approval and the test will run ?

**Comment by janosdebugs** [comment]
> @janosdebugs , there is an option that I can see the result of run here, before you do the review? there is a way to skip on workflows awaiting approval and the test will run ?

Yes, you can run the tests locally (the CI does the same) as described in the [contribution guidelines](https://github.com/oVirt/go-ovirt-client/blob/main/CONTRIBUTING.md).

File: mock_nic_create.go (hunk 15 above)
Comment (janosdebugs) [reply mock_nic_create.go:24]: @mgold1234 this is still not resolved. If I remove this change and run the test suite the test passes. There needs to be a test that fails if this change is not added.

**Review by janosdebugs — CHANGES_REQUESTED** [review]

**Comment by mgold1234** [comment]
@janosdebugs I looked on this example test that also implement negative test:
func TestDiskAttachmentCannotBeAttachedToSecondVM(t *testing.T) {
	t.Parallel()
	helper := getHelper(t)

	vm1 := assertCanCreateVM(
		t,
		helper,
		fmt.Sprintf("disk_attachment_test_%s", helper.GenerateRandomID(5)),
		ovirtclient.CreateVMParams(),
	)
	vm2 := assertCanCreateVM(
		t,
		helper,
		fmt.Sprintf("disk_attachment_test_%s", helper.GenerateRandomID(5)),
		ovirtclient.CreateVMParams(),
	)
	disk := assertCanCreateDisk(t, helper)
	_ = assertCanAttachDisk(t, vm1, disk)
	assertCannotAttachDisk(t, vm2, disk, ovirtclient.EConflict)


this test is not failure, and only catch the specfic error that appear in this case.
I implement the same idea.

**Comment by janosdebugs** [comment]
@mgold1234 please remove the implementation highlighted above and run the test suite. It must fail.

**Comment by janosdebugs** [comment]
@mgold1234 as discussed, we don't have capacity to pursue this edge case further at this time. Closing this PR to revisit #82 after the current release cycle.
